fastchat/serve/monitor/classify/label.py

```diff
@@ -42,94 +49,49 @@ def get_endpoint(endpoint_list):
-            base_url=api_dict["api_base"],
-            api_key=api_dict["api_key"],
-        )
-    else:
-        client = openai.OpenAI()
-
-    output = API_ERROR_OUTPUT
-    for _ in range(API_MAX_RETRY):
-        try:
-            # print(messages)
-            completion = client.chat.completions.create(
-                model=model,
-                messages=messages,
-                temperature=temperature,
-                max_tokens=max_tokens,
-                # extra_body={"guided_choice": GUIDED_CHOICES} if GUIDED_CHOICES else None,
-            )
-            output = completion.choices[0].message.content
-            # print(output)
-            break
-        except openai.RateLimitError as e:
-            print(type(e), e)
-            time.sleep(API_RETRY_SLEEP)
-        except openai.BadRequestError as e:
-            print(messages)
-            print(type(e), e)
-            break
-        except openai.APIConnectionError as e:
-            print(messages)
-            print(type(e), e)
-            time.sleep(API_RETRY_SLEEP)
-        except openai.InternalServerError as e:
-            print(messages)
-            print(type(e), e)
-            time.sleep(API_RETRY_SLEEP)
-        except Exception as e:
-            print(type(e), e)
-            break
-
-    return output
-
-
 def get_answer(
-    question: dict,
+    batch: pd.DataFrame,
     model_name: str,
     max_tokens: int,
     temperature: float,
     answer_file: str,
     api_dict: dict,
-    categories: list,
+    category: object,
     testing: bool,
 ):
-    if "category_tag" in question:
-        category_tag = question["category_tag"]
-    else:
-        category_tag = {}
-
-    output_log = {}
-
-    for category in categories:
-        conv = category.pre_process(question["prompt"])
-        output = chat_completion_openai(
-            model=model_name,
-            messages=conv,
-            temperature=temperature,
-            max_tokens=max_tokens,
-            api_dict=api_dict,
-        )
-        # Dump answers
-        category_tag[category.name_tag] = category.post_process(output)
+    uid_to_row = {}
+    for _, row in batch.iterrows():
+        uid = row["uid"]
+        uid_to_row[uid] = row
+        if "category_tag" in row:
+            TASK_TRACKER[uid].update(row["category_tag"])
+
+    outputs, raw_outputs = category.get_answer(
+        batch, model_name, max_tokens, temperature, api_dict
+    )
+
+    for uid in uid_to_row:
+        output = outputs[uid]
+        TASKS_REQUIRED[uid]["required_tasks"].remove(category.name_tag)
+        TASK_TRACKER[uid][category.name_tag] = output
 
         if testing:
-            output_log[category.name_tag] = output
+            raw_output = raw_outputs[uid]
+            LOGS_TRACKER[uid][category.name_tag] = raw_output
 
-    question["category_tag"] = category_tag
-    if testing:
-        question["output_log"] = output_log
+        row = uid_to_row[uid]
+        if not TASKS_REQUIRED[uid][
```

Comment:
This never gets called if the input data has some of the required tasks completed. For example, I inputed a data file with creative_writing already labeled, and if my config files ask `label.py` to label creative_writing and refusal categories then TASKS_REQUIRED[uid] always look like the following:
```
{'required_tasks': {'creative_writing_v0.1'}}
```

Reply:
I think current implementation would actually just relabel creative_writing again unnecessarily?

Currently, if the output file or cache file already has some labels, that is accounted for when we find the required tasks, but not if the input file already has some labels. I think this was also how it was before I made the changes though